### PR TITLE
Always ask for game path

### LINKED ISSUE #1355: prevent users from using the wrong game path (again)
had a user this week on discord who couldn't run the game
he managed to select vanilla supcom for the base game install instead of forged alliance

please don't allow users to select the vanilla supcom install directory (tested myself and the client doesn't throw an error when trying to do so in 0.10.7)

### LINKED ISSUE #1594: Player not always asked if he has not set the game path
**Wanna have the bug fixed quickly?**
Visit Issue hunt...
[Issue hunt](https://issuehunt.io/r/FAForever/downlords-faf-client)

### PR DESCRIPTION
**Pull request by 1-alex98**
#1594

### CODE REVIEW

File: src/main/java/com/faforever/client/game/GameService.java [hunk 1]
```diff
@@ -312,6 +319,12 @@ public ReadOnlyBooleanProperty gameRunningProperty() {
       return completedFuture(null);
     }
 
+    if (!preferencesService.isGamePathValid()) {
+      CompletableFuture<Path> gameDirectoryFuture = new CompletableFuture<>();
+      eventBus.post(new GameDirectoryChooseEvent(gameDirectoryFuture));
+      return gameDirectoryFuture.thenCompose(path -> joinGame(game, password));
+    }
+
```
Comment (Geosearchef) [inline src/main/java/com/faforever/client/game/GameService.java]: This code is duplicated 6 times, I'd extract the first 2 lines to a method returning the gameDirectoryFuture.
```
if (!preferencesService.isGamePathValid()) {
    return ensureGameDirectorySet().thenAccept(p -> joinGame()/hostGame/whatever);
}
```

**Review by Geosearchef — DISMISSED** [review]

File: src/main/java/com/faforever/client/game/GameService.java (hunk 1 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/game/GameService.java]: true

Comment (1-alex98) [reply src/main/java/com/faforever/client/game/GameService.java]: Acctually no it depends on the method look closely to the differences

Comment (1-alex98) [reply src/main/java/com/faforever/client/game/GameService.java]: There is not much to refactor

**Comment by Askaholic** [comment]
While you are working on this can you take a look at #1100 and #1355. Those problems may be solvable or solved by similar changes.

**Comment by 1-alex98** [comment]
Well they are related yes but seperate issues

**Comment by 1-alex98** [comment]
#1100 kind of gets fixed by that because it checks with every game start which kind of is periodically

**Comment by 1-alex98** [comment]
Before the check was only done sometimes depending on how u started the game

**Comment by 1-alex98** [comment]
The check is also triggered if the game path is set but invalid

**Comment by 1-alex98** [comment]
Also fixes #1355

**Comment by Slothologist** [comment]
The hashes created by the "sha256OfFile" function differ from the actual hashes (at least on linux). 
Because of that I never run into the "vanilla exe detected" case.
I fixed that in https://github.com/FAForever/downlords-faf-client/compare/bugfix/%231594-ask-for-game...Slothologist:bugfix/%231594-ask-for-game

It would maybe be better to check more explicitly like
```
If calculatedHash == FA.exe
    yay
else if calculatedHash == vanilla.exe
    oh no, you selected the wrong folder
else
    wtf this should not happen
```

**Comment by 1-alex98** [comment]
Well there are a ton of different FA version that are supported / ok I think so getting all the hashes might be hard @Slothologist I am not an expert on this but the steam version differs from the retail one but both are supported and therefore might have different hashes and who knows which other retail versions exist out there

**Comment by 1-alex98** [comment]
@Slothologist are u on Zulip? You wanna become part of Zulip(chat service)? We somehow need to test your thing on windows as well right?

**Comment by 1-alex98** [comment]
Write me an email if u want to alexander.von.trostorff@gmail.com

**Comment by axel1200** [comment]
does not close for me

**Comment by Slothologist** [comment]
> @Slothologist are u on Zulip? You wanna become part of Zulip(chat service)? We somehow need to test your thing on windows as well right?

Thanks for the Zulip invite, I am now part of that.
Furthermore, I tried your version out in Windows and the hash the function provides is identical to the one I got on Linux. So something like xng1gM.... for the FA exe (which is wrong). 

I see the problem with acquiring the hashes of all possible versions, but atm we are only filtering for the steam vanilla version.
Maybe we could just print a warning if an unknown exe is presented? Like "Alright, we seem to have no knowledge about this exe, make sure that you have selected the right folder, but go ahead, we won't stop you"?

**Comment by 1-alex98** [comment]
I think the way it is now is good(in your branch). For most ppl it will do and it is also a clear improvement. I rather have zero false positive errors or warnings and stay with it like it is now.

**Comment by 1-alex98** [comment]
Merged in your changes @Slothologist

**Comment by 1-alex98** [comment]
tested again works out well

**Comment by 1-alex98** [comment]
The hash is defined in the properties file and it is a list should there ever be other hashes we can simply add them by adding one line. So everything covered .;)

**Pull request merged**